boa_gc/src/lib.rs

```diff
@@ -1,6 +1,375 @@
-    GcCellRef as Ref, GcCellRefMut as RefMut, Trace,
-};
+#![warn(
+    clippy::perf,
+    clippy::single_match_else,
+    clippy::dbg_macro,
+    clippy::doc_markdown,
+    clippy::wildcard_imports,
+    clippy::struct_excessive_bools,
+    clippy::doc_markdown,
+    clippy::semicolon_if_nothing_returned,
+    clippy::pedantic
+)]
+#![deny(
+    clippy::all,
+    clippy::cast_lossless,
+    clippy::redundant_closure_for_method_calls,
+    clippy::unnested_or_patterns,
+    clippy::trivially_copy_pass_by_ref,
+    clippy::needless_pass_by_value,
+    clippy::match_wildcard_for_single_variants,
+    clippy::map_unwrap_or,
+    clippy::undocumented_unsafe_blocks,
+    clippy::missing_safety_doc,
+    unsafe_op_in_unsafe_fn,
+    unused_qualifications,
+    unused_import_braces,
+    unused_lifetimes,
+    unreachable_pub,
+    trivial_numeric_casts,
+    rustdoc::broken_intra_doc_links,
+    missing_debug_implementations,
+    missing_copy_implementations,
+    deprecated_in_future,
+    meta_variable_misuse,
+    non_ascii_idents,
+    rust_2018_compatibility,
+    rust_2018_idioms,
+    future_incompatible,
+    nonstandard_style,
+    missing_docs
+)]
+#![allow(clippy::let_unit_value, clippy::module_name_repetitions)]
+
+extern crate self as boa_gc;
+
+use boa_profiler::Profiler;
+use std::cell::{Cell, RefCell};
+use std::mem;
+use std::ptr::NonNull;
+
+mod trace;
+
+pub(crate) mod internals;
+
+mod cell;
+mod pointers;
+
+pub use crate::trace::{Finalize, Trace};
+pub use boa_macros::{Finalize, Trace};
+pub use cell::{GcCell, GcCellRef, GcCellRefMut};
+pub use pointers::{Ephemeron, Gc, WeakGc};
+
+use internals::GcBox;
+
+type GcPointer = NonNull<GcBox<dyn Trace>>;
+
+thread_local!(static EPHEMERON_QUEUE: Cell<Option<Vec<GcPointer>>> = Cell::new(None));
+thread_local!(static GC_DROPPING: Cell<bool> = Cell::new(false));
+thread_local!(static BOA_GC: RefCell<BoaGc> = RefCell::new( BoaGc {
+    config: GcConfig::default(),
+    runtime: GcRuntimeData::default(),
+    adult_start: Cell::new(None),
+}));
+
+#[derive(Debug, Clone, Copy)]
+struct GcConfig {
+    threshold: usize,
+    used_space_percentage: usize,
+}
+
+// Setting the defaults to an arbitrary value currently.
+//
+// TODO: Add a configure later
+impl Default for GcConfig {
+    fn default() -> Self {
+        Self {
+            threshold: 1024,
+            used_space_percentage: 80,
+        }
+    }
+}
+
+#[derive(Default)]
+struct GcRuntimeData {
+    collections: usize,
+    bytes_allocated: usize,
+}
+
+struct BoaGc {
+    config: GcConfig,
+    runtime: GcRuntimeData,
+    adult_start: Cell<Option<GcPointer>>,
+}
+
+impl Drop for BoaGc {
+    fn drop(&mut self) {
+        Collector::dump(self);
+    }
+}
+// Whether or not the thread is currently in the sweep phase of garbage collection.
+// During this phase, attempts to dereference a `Gc<T>` pointer will trigger a panic.
+
+struct DropGuard;
+
+impl DropGuard {
+    fn new() -> DropGuard {
```

Comment:
This is missing some docs, with information on the default value. Also, it can return Self, to make it easier to rename.